src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h

```diff
@@ -16,85 +16,99 @@
- * @tparam T base precision for most computation
- * @tparam T_FP high precision for matrix inversion, T_FP >= T
+ * @tparam VALUE base precision for most computation
+ * @tparam VALUE_FP high precision for matrix inversion, T_FP >= T
  */
-template<typename T, typename T_FP>
+template<typename VALUE, typename VALUE_FP>
 class MatrixUpdateOMPTarget
 {
 public:
-  using WFT = WaveFunctionTypes<T, T_FP>;
-  using This_t = MatrixUpdateOMPTarget<T, T_FP>;
+  using WFT           = WaveFunctionTypes<VALUE, VALUE_FP>;
+  using Value         = typename WFT::Value;
+  using FullPrecValue = typename WFT::FullPrecValue;
+  using LogValue      = typename WFT::LogValue;
+  using This_t        = MatrixUpdateOMPTarget<VALUE, VALUE_FP>;
+
+  // There is no expectation that this OMP only code can possibly work with anything but OMPallocator based
+  // Allocators.  So this file should never include DualAllocatorAliases.hpp this violation of YAGNI isn't.
+  // I do this for symmetry with MatrixDelayedUpdateCUDA only.
   template<typename DT>
   using PinnedDualAllocator = OffloadPinnedAllocator<DT>;
 
-  using OffloadValueVector_t       = Vector<T, OffloadAllocator<T>>;
-  using OffloadPinnedValueVector_t = Vector<T, OffloadPinnedAllocator<T>>;
-  using OffloadPinnedValueMatrix_t = Matrix<T, OffloadPinnedAllocator<T>>;
+  template<typename DT>
+  using UnpinnedOffloadVector = Vector<DT, OffloadAllocator<DT>>;
+  template<typename DT>
+  using OffloadVector = Vector<DT, OffloadPinnedAllocator<DT>>;
+  template<typename DT>
+  using OffloadMatrix = Matrix<DT, OffloadPinnedAllocator<DT>>;
+
+  struct MatrixUpdateOMPTargetMultiWalkerMem : public Resource
+  {
+    // constant array value T(1)
+    UnpinnedOffloadVector<Value> cone_vec;
+    // constant array value T(0)
+    UnpinnedOffloadVector<Value> czero_vec;
+    // multi walker of grads for transfer needs.
+    OffloadMatrix<Value> grads_value_v;
+    // pointer buffer
+    Vector<char, PinnedAllocator<char>> buffer_H2D;
+    /// scratch space for rank-1 update
+    UnpinnedOffloadVector<Value> mw_temp;
+    // scratch space for keeping one row of Ainv
+    UnpinnedOffloadVector<Value> mw_rcopy;
+
+
+    MatrixUpdateOMPTargetMultiWalkerMem() : Resource("MatrixUpdateOMPTargetMultiWalkerMem") {}
+
+    MatrixUpdateOMPTargetMultiWalkerMem(const MatrixUpdateOMPTargetMultiWalkerMem&)
+        : MatrixUpdateOMPTargetMultiWalkerMem()
+    {}
+
+    Resource* makeClone() const override { return new MatrixUpdateOMPTargetMultiWalkerMem(*this); }
+  };
 
   /// matrix inversion engine
-  DiracMatrix<T_FP> detEng;
+  DiracMatrix<VALUE_FP> detEng;
+
+  int inv_row_id_;
+
+  /** matrix inversion engine this crowd scope resouce and only the leader engine gets it.
+   *  With some care this could be the DiracMatrixComputeCUDA 
+   *  as well as DiracMatrixComputeOMPTarget
+   */
+  using DetInverter = DiracMatrixComputeOMPTarget<VALUE_FP>;
+  UPtr<DetInverter> det_inverter_;
```

Comment:
This doesn't seem to belong to this PR and I cannot find det_inverter_ being used.

Reply:
It is needed for now.